pinot-query-runtime/src/main/java/org/apache/pinot/query/runtime/plan/PhysicalPlanVisitor.java

```diff
@@ -61,12 +63,20 @@ public static OpChain build(StageNode node, PlanRequestContext context) {
 
   @Override
   public MultiStageOperator visitMailboxReceive(MailboxReceiveNode node, PlanRequestContext context) {
-    MailboxReceiveOperator mailboxReceiveOperator =
-        new MailboxReceiveOperator(context.getOpChainExecutionContext(), node.getExchangeType(),
-            node.getCollationKeys(), node.getCollationDirections(), node.isSortOnSender(), node.isSortOnReceiver(),
-            node.getDataSchema(), node.getSenderStageId(), node.getStageId());
-    context.addReceivingMailboxes(mailboxReceiveOperator.getSendingMailbox());
-    return mailboxReceiveOperator;
+    if (!CollectionUtils.isEmpty(node.getCollationKeys()) && node.isSortOnReceiver()) {
```

Comment:
Is there a case where `node.isSortOnReceiver()` is true but collation keys are empty?

Reply:
Collation keys should not be empty to use `node.isSortOnReceiver()`, but the opposite is possible. We can have collation keys as not empty, but the `isSortOnReceiver` is false. In this case I'll be creating a normal `MailboxSendOperator` instead of the sorted version. This is useful for scenarios where the data is partitioned and sorted on the sender in the future, in which case the receiver sort is a no-op.

Reply:
Yeah makes sense.

Should we change this condition to simply: `node.isSortOnReceiver()` and add a precondition under the if block? Might be helpful to catch bugs.

Reply:
oh wait, actually today for ORDER BY queries I always set 'isSortOnReceiver' to true. I've modified that to explicitly set based on whether collation keys exist or not. let me know if this seems okay or not. cc @walterddr

Reply:
I am not following the discussion on this one --> if `node.isSortOnReceiver() == true`, do we still generate the SortNode on top of the receive?

Reply:
So now with my changes the following will hold:

- `node.isSortOnReceiver() == true`: only set if actual sorting needs to be performed on the receiver
    - Window functions that need ORDER BY
    - ORDER BY queries that have at least one collation key
-  'node.isSortOnReceiver() == false`: can be false if collation key is non-empty (we don't use this code path today) such as when in the future we have partitioned data and decide to sort on sender side only.
    - Window functions without ORDER BY
    - ORDER BY queries with no collation key but with offset and limit

The `SortNode` will still be created where needed, just if it's child operator is sorted then we set that the input is already sorted, otherwise we set input is not sorted (just as before). This discussion is more around when to create the sorted mailbox receive vs. non-sorted one

Reply:
> So now with my changes the following will hold:
> 
> * `node.isSortOnReceiver() == true`: only set if actual sorting needs to be performed on the receiver
>   
>   * Window functions that need ORDER BY
>   * ORDER BY queries that have at least one collation key
> * `node.isSortOnReceiver() == false`: set when sorting not required on receiver side.
>   * Can be `false` if collation key is non-empty (we don't use this code path today) such as when in the future we have partitioned data and decide to sort on sender side only.

[RR] hmm. could you explain more as: under what circumstances will this be true? did you mean there's an "ORDER BY " clause in the SQL, or there is actually a collation key specified on the SortExchange RelNode after Calcite has finish optimizer phase?

> 
> The `SortNode` will still be created where needed, just if it's child operator is sorted then we set that the input is already sorted, otherwise we set input is not sorted (just as before). This discussion is more around when to create the sorted mailbox receive vs. non-sorted one

[RR] the above-described behavior is identical to prior to https://github.com/apache/pinot/pull/10408 yes? just instead of doing the sorting in `SortExchange`, previously we will create a SortNode on top of the Exchange Node?

Reply:
> [RR] hmm. could you explain more as: under what circumstances will this be true? did you mean there's an "ORDER BY " clause in the SQL, or there is actually a collation key specified on the SortExchange RelNode after Calcite has finish optimizer phase?

I meant if there is a collation key specified in the SortExchange RelNode after the Calcite optimizer phase. I was just giving an example of ORDER BY queries

> [RR] the above-described behavior is identical to prior to https://github.com/apache/pinot/pull/10408 yes? just instead of doing the sorting in SortExchange, previously we will create a SortNode on top of the Exchange Node?

yes the end behavior is the exact same In terms of creation of the SortNode. Earlier the plan indicated sortOnReceiver as true for all LogicalSortExchange created for SortNodes without checking if collation key is set, but we handled skipping sorting in the receive operator. Now we explicitly set that sortOnReceiver is false (see planner test changes). Does that help clarify?

Reply:
The node creation is the exact same as before, it’s just about setting the isSortOnReceiver flag

Reply:
Discussed this offline with @walterddr and we are in agreement about this change and discussion